writers.writePython2: remove / writers.writePyPy{2,3}: init

wamserma opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

https://github.com/NixOS/nixpkgs/issues/148779
python2 is EOL and due for removal; functions unused in nixpkgs

a replacement using the PyPy interpreter is added to ease transition
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 150132"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

wamserma committed:
writers.writePython2: remove

rnhmjoj reviewed (changes requested):
This change needs an entry in the release notes and maybe an alias.

Lassulus commented:
maybe we should inline the makePythonWriter function into writePython3?

wamserma committed:
writers.PyPy{2,3}: init

wamserma commented:
> This change needs an entry in the release notes and maybe an alias.

PR updated, also added writers for PyPy as an alternative for legacy scripts.

rnhmjoj commented:
> PR updated, also added writers for PyPy as an alternative for legacy scripts.

Thanks you. Unfortunately the aliases seem to be messing with `pkgs.tests.writers` somehow. It has probably to do with `writers` being an attrset, but I don't know how to make an alias in this case.

wamserma commented:
> > PR updated, also added writers for PyPy as an alternative for legacy scripts.
> 
> Thanks you. Unfortunately the aliases seem to be messing with `pkgs.tests.writers` somehow. It has probably to do with `writers` being an attrset, but I don't know how to make an alias in this case.

Ah yes, fixed.

Lassulus commented:
sadly this has already a merge conflict.

wamserma commented:
> sadly this has already a merge conflict.

Rebased the PR.

Lassulus commented:
hmm, seems like the pypy3 writer test doesn't work for me:
https://p.krebsco.de/0bww66n

wamserma commented:
> hmm, seems like the pypy3 writer test doesn't work for me

That is due to `pypy3Packages.bootstrapped-pip` not building. Nothing caused by this PR.

Profpatsch commented:
Why? Why remove it? Does that mean all programs written for python 2 have to be deleted now, because somebody declared it “EOL”?

Profpatsch commented:
I’d like to revert this until python2 is actually removed (iff that ever happens which I honestly hope it won’t).

Profpatsch commented:
To be clear, I am rather angry at this change, breaking an API like this without any good reason (and no, “EOL” is *not* a good reason!) is a total no-go.

rnhmjoj commented:
> and no, “EOL” is not a good reason!

It totally is for me: the CPython 2 interpreter and its standard library almost surely have security vulnerabilities that will never be fixed.

7c6f434c commented:
>> and no, “EOL” is not a good reason!
>
>It totally is for me: the CPython 2 interpreter and its standard library almost surely have security vulnerabilities that will never be fixed.

Good question what kind of data writePython scripts actually handle.

But even without looking into that, that's EOL for CPython2, not Python2. We have an upstream supported Python2 interpreter in Nixpkgs.

wamserma commented:
@Profpatsch This is just a helper to write a python script from a nix expression and have a given interpreter (Python2) set. To my knowledge this helper was not used in nixpkgs and most likely unused outside nixpkgs.
We want to retire Python2 from nixpkgs, because it is EOL and will not get any security fixes. 
If you still need to write scripts with Python2 syntax from a nix expression, you can simply switch to writePyPy2, which will set PyPy as interpreter in the scripts shebang line. All other are urged to convert their scripts to Python3.
Further, we did not "break" the API, we rather deprecated it. You can always revive the python2-writer with an overlay without side-effects on the other script-writing helpers, if you have the need to keep it.
